Double Entry Accounting

rafabap opened the pull request:
Hybrid accounting system that uses inventories and double-entry accounting.

The Double-entry is simplified so that it uses only stock accounts (ASSETS, LIABILITIES, EQUITY) and ignores flow accounts (INCOME, EXPENSES). Each account has an inventory of contracts.

Each account can hold only one type of contract in its inventory.

The Agents interact with their accounts via a Book, that provides a series of operations given in BookAPI.

rafabap committed:
Added accounting files

rafabap committed:
One account <=> One Contract Type

Made changes to ensure a one-to-one mapping between accounts and contract types

rafabap committed:
Delete Ledger.java

rafabap committed:
Delete Collateral.java

davidrpugh commented:
@rafabap Let's start with the `Book` interface.  First, I do not understand the terminology. A `Book` seems to be a collection of `Account` (or `Contract`) objects in which case why not call this a `BalanceSheet`? Second if a `Book` is just a collection of contracts then I think we should only allow adding/removing of contracts from the `Book` as well as valuing assets liabilities and equity.  Logic for updating the value of a particular contract should be pushed elsewhere.

Excuse the Scala, but something like the following...

```scala
abstract class Contract {
  
  /** agent for whom the contract is a liability */
  def issuer: Agent

  /** agent for whom the contract is an asset */
  def counterparty: Agent

 /** Objective value of a contract */
  def faceValue: Double

} 


/** Immutable storage container for a particular agent's contracts. */
class BalanceSheet[+C <: Contract] private(agent: UUID, contracts: GenSet[C] ) {

  /** Returns a new BalanceSheet with an additional contract .*/
  def add[C1 >: C](contract: C1): BalanceSheet[C1] = {
    new BalanceSheet(agent, contracts + contract)  
  }

  /** Returns a new BalanceSheet after removing a particular contact. */
  def remove(contract: C): BalanceSheet[C] = {
    new BalanceSheet(agent, contracts - contract)
  } 

  /** Collection of contracts that are assets. */
  def assets: GenSet[C] = {
    contracts.filter(contract => contract.counterparty == agent)
  }

  /** Collection of contracts that are liabilities. */
  def liabilities: GenSet[C] = {
    contracts.filter(contract => contract.issuer == agent)
  }

  /** The value of equity is just the difference between the value of assets and the value of liabilities. 
     * @note the value of equity is subjective and depends on the value function `v` used to value the 
     * underlying contracts.
     */
  def equity(v: (Contract) => Double): Double = {
    assets.map(v).sum - liabilities.map(v).sum
  }

  /** Leverage ratio */
  def leverage(v: (Contract) => Double): Double = {
    assets.map(v).sum / equity(v)
  }
  
}
```

davidrpugh commented:
@rafabap I am also confused about the difference between `Contract` and an `Account`.

davidrpugh commented:
My most general comment is that I find the terminology very confusing.  I generally do not think of 'Asset', 'Liability' and `Equity` as being types of `Accounts`.  But might just be me.  

I guess you could have something like...

```scala
class DepositAccount private(issuer: UUID, counterparty: UUID, initialDeposit: Double) extends Contract {
  val faceValue = initialDeposit

  def debit(value: Double): DepositAccount = {
    new DepositAccount(issuer, counterparty, initialDeposit - value)
  } 

  def credit(value: Double): DepositAccount = {
    new DepositAccount(issuer, counterparty, initialDeposit + value)
  }

}

val agent: UUID = ???
val bank: UUID = ???

val deposit = DepositContract(bank, agent, 1e6)

```
in the above

davidrpugh commented:
@rafabap I just realized that `Book` used to be called `Ledger`!  I much prefer `Ledger` as it makes good use of standard accounting terminology (which has been picked up by the blockchain community which I think is useful).

davidrpugh commented:
@rafabap Some specific comments about the `Account` class.  

* Is the account name supposed to uniquely identify the account? If so, then perhaps better to use `UUID` then `String` for the type.  
* Why use an `Enum` to model different account types? Is the `Enum` values just to allow you to switch behavior depending on account type?
* Where does the `Action` type imported from? Is this a Java type? Or has this ours?

davidrpugh commented:
@rafabap Perhaps my lack of comfort with your terminology stems from my being used to Perry Mehrling's  terminology.  I am reading through the [Double-Entry Book-Keeping](https://en.wikipedia.org/wiki/Double-entry_bookkeeping_system) Wikipedia entry to try and familiarize myself with your terms. Is this the best source?

rafabap commented:
@davidrpugh sorry for the confusion! Just a few pints to clarify:

1) double-entry
The Wikipedia article is one of the best resources I've found. The problem is that although the rules are consistent across sources, the terminology isn't!
I'm happy to go back to using Ledger rather than Book. 

The types of Account only matter when defining the sign of debits and credits.

2) double entry vs Inventories

I see our accounting system as a hybrid composed of two parts. The first is the double entry rules, which requires separate accounts implementing their debits and credits. The second is the inventory: a way to keep lists of contracts. 

My solution is to combine the two by starting with a standard double-entry ledger havjng a few accounts, and then rather than having a single inventory for the agent, I split it in smaller inventories, each associated to one account. In addition, for clarity, I force that each account includes only one type of contract in its inventory.

prauwolf commented on Account.java:
```diff
@@ -0,0 +1,86 @@
+
+    private double balance;
+
+//    private Collateral collateralType;
```
remove commented variables, unless it is a discussion point whether or not to add them.

prauwolf commented on Account.java:
```diff
@@ -0,0 +1,86 @@
+    protected HashSet<Contract> contracts;
+
+
+    void addContract(Contract contract) {
```
is it intentional that these functions scope are the Java default? It's likely that's best, but I didn't know if some, like addContract, should be public?

prauwolf commented on Account.java:
```diff
@@ -0,0 +1,86 @@
+        return availableActions;
+    }
+
+//    public void setCollateralType(Collateral collateralType) {
```
remove?

prauwolf commented on Account.java:
```diff
@@ -0,0 +1,86 @@
+    String getName() {
+        return name;
+    }
+}
```
Should there be a removeContract functionality?

prauwolf replied:
Should the user be able to pull a list of the existing contracts in the contract hashSet?

prauwolf commented on BookAPI.java:
```diff
@@ -0,0 +1,27 @@
+    void addCash(double amount);
+
+
+    void sellAsset(double amount, Class<? extends Contract> assetType);
```
is 'sellAsset and 'payLiability' the only way to remove a contract? Is that what we want? It very well maybe.

prauwolf commented on Book.java:
```diff
@@ -0,0 +1,315 @@
+        assetAccountsMap = new HashMap<>();
+        liabilityAccountsMap = new HashMap<>();
+
+        // Not sure whether I should be passing the owner...
```
I think this is a good idea. There may be times when the book realizes that something requires the owner's attention. With this architecture, the book can ping the owner if required.

prauwolf commented on Book.java:
```diff
@@ -0,0 +1,315 @@
+        // These hashmaps are used to access the correct account for a given type of contract.
+        // Note that separate hashmaps are needed for asset accounts and liability accounts: the same contract
+        // type (such as Loan) can sometimes be an asset and sometimes a liability.
+        assetAccountsMap = new HashMap<>();
```
I'm a little confused about this, perhaps we can discuss it in person.

bherd replied:
General question: is thread safety an issue at this point in time? If yes, then this would require some sort of synchronisation, because HashMaps and HashSets are (as far as I'm aware) not thread-safe.

prauwolf commented on Book.java:
```diff
@@ -0,0 +1,315 @@
+ *
+ * @author rafa
+ */
+public class Book implements BookAPI {
```
Is 'Book' intended to be a parent class to most Books, or an example of using the BookAPI?

prauwolf commented on Book.java:
```diff
@@ -0,0 +1,315 @@
+        this.owner = owner;
+
+        // A book is initially created with a cash account and an equityAccounts account (it's the simplest possible book)
+        cashAccount = new Account("cash", AccountType.ASSET);
```
related to my previous question regarding whether 'Book' is the parent class of most other Books. Perhaps the constructor should contain parameters for the types of accounts which should be initialized?

prauwolf commented on Book.java:
```diff
@@ -0,0 +1,315 @@
+        return equityTotal;
+    }
+
+    public double getAssetValueOf(Class<? extends Contract> contractType) {
```
I'm a little confused. How are contracts being valued? Can anyone send in a subjective function for evaluation?

bherd replied:
Yes, that sounds like a good place for a higher-order function that accepts a user-defined valuation function as a parameter and applies it to the given contract.

prauwolf commented on Book.java:
```diff
@@ -0,0 +1,315 @@
+        switch(account.getAccountType()) {
+
+            case ASSET:
+                assetAccounts.add(account);
```
will this overwrite existing accounts of the same type?

prauwolf commented on Book.java:
```diff
@@ -0,0 +1,315 @@
+            case EQUITY:
+                equityAccounts.add(account);
+
+            // Not sure what to do with INCOME, EXPENSES
```
Just a reminder to discuss this comment during the review

prauwolf commented on Book.java:
```diff
@@ -0,0 +1,315 @@
+        equityAccounts = new HashSet<>();
+
+        // Each Account includes an inventory to hold one type of contract.
+        // These hashmaps are used to access the correct account for a given type of contract.
```
Why does each account only hold one type of contract? What if someone sets it up to hold the most generic level of contract?

prauwolf commented on Book.java:
```diff
@@ -0,0 +1,315 @@
+     * @param contract an Asset contract to add
+     */
+    public void addAsset(Contract contract) {
+        Account assetAccount = assetAccountsMap.get(contract.getClass());
```
Should there be a check that the contract is an asset? I'm not sure how one tells, but could the contract be miscategorized?

prauwolf replied:
This will check the lowest level of contract (i.e. not the contract's super class), and create an account of that type. Could an account be of a super class of a Contract?

prauwolf commented on Book.java:
```diff
@@ -0,0 +1,315 @@
+        }
+
+        // (dr asset, cr equity)
+        assetAccount.debit(contract.getValue());
```
Do contracts have a 'getValue()'? Perhaps they should, if they don't. We should discuss contract valuation and how it relates to this system.

bherd replied:
I'm not sure if valuation shouldn't be external to the contract, e.g. by means of a function that accepts a contract and returns a value. If it is internal to the contract, then it will be difficult to have different types of valuation. Also, if valuation is done in a more functional way, it will become easier to plug together more complex types of valuation. But that's probably a separate discussion.

prauwolf commented on Book.java:
```diff
@@ -0,0 +1,315 @@
+        Account liabilityAccount = liabilityAccountsMap.get(liabilityType);
+
+        //Todo: What do we do if we can't pay??!! At the moment I'm calling my owner to raise liquidity
+        if (cashAccount.getBalance() < amount) {
```
We should probably decide how we're going to deal with errors in the library.

davidrpugh replied:
@prauwolf Agreed.  I don't know much about the Java conventions regarding checked vs unchecked exceptions, etc.  I think we need some kind of `LiquidityException` that will be thrown in the event that an agent does not have liquid resources available to make required payments.  Ideally this would also be expressed in the return type of the method (which would force the caller of the method to explicitly deal with the possibility that an exception would be returned.

prauwolf commented on Book.java:
```diff
@@ -0,0 +1,315 @@
+     * If I've sold an asset, debit cash and credit asset
+     * @param amount the *value* of the asset
+     */
+    public void sellAsset(double amount, Class<? extends Contract> assetType) {
```
As implemented, what class should be calling 'sellAsset'?

prauwolf commented on Book.java:
```diff
@@ -0,0 +1,315 @@
+     * @param me the owner of the Book
+     * @return an ArrayList of Actions that are available to me at this moment
+     */
+    public ArrayList<Action> getAvailableActions(Agent me) {
```
Let's discuss actions in the code review. I really like the idea. Perhaps we should have some kind of interface which defines a list of actions, since we'd probably use them in a variety of contexts.

prauwolf commented on Book.java:
```diff
@@ -0,0 +1,315 @@
+
+        for (Contract contract : assetAccount.contracts) {
+            Asset asset = (Asset) contract;
+            if (asset.priceFell()) {
```
Can you explain this in a bit more detail in the review?

bherd commented on Book.java:
```diff
@@ -0,0 +1,315 @@
+        assetAccounts = new HashSet<>();
+        liabilityAccounts = new HashSet<>();
+        equityAccounts = new HashSet<>();
+
```
Is it strictly necessary to have different hash sets holding the same account objects? Although (I think) Java stores values by reference, I fear there might be situations where, for some reason, the account object gets copied and thus ends up being 'forked'.  

Would it be better to use just one single data structure `accounts` and use clever access functions to get and set the type of content that is relevant? For example, to return all accounts from `accounts` in a lazy way (i.e. as a stream), you could define the following function (sorry, haven't checked if it compiles):

```
Stream<Object> getAccounts() {
  return accounts.entrySet().stream.map(Map.Entry::getValue);
}
```

To return all accounts of a particular type, you could define the following function:

```
Stream<Object> getAccountsOfType(AccountType t) {
  return accounts.entrySet().stream().filter(entry -> entry.getKey() == t).map(Map.Entry::getValue);
}
```

In order to turn the streams into lists, you could also define a general function:

```
List<?> toList(Stream<Object> s) {
  return s.collect(Collectors.toList());
}
```

The same could, of course be done in the opposite direction, i.e. to set objects in the map. You could then get rid of the other data structures and only keep a single source of truth. 

As I said, I haven't checked whether the snippets compile, so please take them with a pinch of salt :)

davidrpugh commented:
@rafabap and @bherd Based on our discussion today we decided to move forward with a message passing solution for handling interaction between `Ledger` / `Book` and `Transactions`, etc.  I think we should take a hard look at [Akka](http://akka.io/).  Existence of widely used [Java](http://doc.akka.io/docs/akka/2.4/java.html?_ga=1.187894176.96835854.1486449084) and [Scala](http://doc.akka.io/docs/akka/2.4/scala.html?_ga=1.154463024.96835854.1486449084) APIs is nice and using the Actor model as our concurrency solution will help constrain our design space in general...

rafabap committed:
Private debit/credit, created doubleEntry function

Made a static, legal `doubleEntry(Account debitAccount, Account creditAccount, double amount)` function

rafabap committed:
Implemented changes from code review

A `Ledger` now has:
- a set of `Account`s
- two sets of `Contract`s to act as inventory (one for assets and one for liabilities)

By splitting the inventory from the accounts, I have avoided issues to do with searching of the inventories etc. The invariants maintained: 
- each `Account` corresponds to one type of `Contract`, and the sum of the value of the contracts of that type equals the balance of the account. This invariant is not hard-coded but will be.
- Assets = Liabilities + Equity (guaranteed by the double entry)

rafabap commented:
Hi all, I have implemented most of the comments received so far and during the code review. The concurrent implementation of the `Ledger` is left for a future date.

I have followed David's advice and decided to split the inventories from the Accounts. Therefore, a `Ledger` has, on the one hand, a set of accounts; and on the other hand, an inventory of contracts (the inventory is split into assets and liabilities for convenience).

Searching etc. of the inventory is done using functional calls as suggested by Ben.

The invariants provided by the framework are:

- Assets = Liabilities + Equity, as guaranteed by the double entry
- For each type of Contract, there exists an Account associated with it. The invariant is that the sum of the value of all contracts of that type is equal to the balance of the account. This invariant is not hard-coded yet.

rafabap committed:
Deleted Book

rafabap committed:
Deleted BookAPI
